src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java

```diff
@@ -205,14 +206,15 @@ public void onClose(int i, String s) {
                         // The workspace is not known in advance, so we have to execute a cd command.
                         watch.getInput().write(
                                 String.format("cd \"%s\"%s", pwd, NEWLINE).getBytes(StandardCharsets.UTF_8));
+
                     }
 
                     if (environmentExpander != null) {
                         EnvVars envVars = new EnvVars();
                         environmentExpander.expand(envVars);
                         for (Map.Entry<String, String> entry : envVars.entrySet()) {
                             watch.getInput().write(
-                                    String.format("export %s=\"%s\"%s", entry.getKey(), entry.getValue(), NEWLINE).getBytes(StandardCharsets.UTF_8));
+                                    String.format("export %s=\"%s\"%s", entry.getKey(),  StringEscapeUtils.escapeJavaScript(entry.getValue()), NEWLINE).getBytes(StandardCharsets.UTF_8));
```

Comment:
This escapes single quotes as well, which

a) does not even work in bash
b) is wrong as we are already in double quotes (`%s=\"%s\"%s"`)

see e.g. a password that contains a single quote (stupid idea, but happens), where it will be encoded as `\'` (literally).

example pipeline:

```
podTemplate(name: "horst", label: "label", containers: [
                        containerTemplate(name: 'ubuntu',
                                image: 'ubuntu',
                                ttyEnabled: true,
                                command: 'cat',
                        )
]) {
    withEnv(["C='C'", 'D=\'', 'E=	', 'F=\n']) {
        node("label") {
            container('ubuntu') {
                sh '''#!/bin/bash -x
ENC='Jw=='; [[ $D = $(base64 -d <<<$ENC) ]]; echo $?
            }
        }
    }
}
```

The single quote part can be fixed by using `escapeJava` instead of `escapeJavaScript`, but I don't know if that will have more issues. It also escapes "quotes and control-chars (tab, backslash, cr, ff, etc.)", see Javadoc. It also escapes unicode characters, see source code :-).

Escaping is hard, especially for bash. Maybe it makes sense to add more tests, especially regarding control chars and unicode.

Reply:
Back now and able to look at this - I had originally thought of the issue of blasting these out to build console log and thought to use the javascript to escape - how wrong that was!

re: xss - I check and the console log viewer escapes already
```
export HACK="<script>alert('XSS);</script>"
```
becomes
```
export HACK="&lt;script&gt;alert('XSS');&lt;\/script&gt;" 
```
so I think we should be good there unless I'm not seeing it right. 

To sum up: We want to only escape double quotes if they exist in the environment variable, not anything else with it - otherwise, we're guessing at what the user REALLY meant to escape. If the user provided an escaped quote such as 
```
  environment {
   //note that pipeline requires that extra \
    quote = 'Testing-\\"-qoute' 
    }
```
it comes to bash as 
`quote = 'Testing-"-qoute' `
and we should provide the same functionality as the other executors provide by putting printing `export quote = "Testing-\"-qoute"` 
and seeing 
`quote = 'Testing-"-qoute' `

So I'm not sure where to go from here other than making naive assumption to ONLY escape a double quote in the provided variable then let bash double quotes handle the rest.

Reply:
XSS is not an issue here, I never wanted to hint at that. OWASP is just a nice resource for escaping issues of any kind 😃 .

Reply:
> So I'm not sure where to go from here

I would suggest to use `escapeJava` instead of `escapeJavascript`, and add a few test cases to make sure that whitespace characters (tab, backslash, cr, ff, etc.) are encoded correctly. See https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringEscapeUtils.html#escapeJava(java.lang.String) for reference.